resources/migrations/001_update_migrations.yaml

```diff
@@ -7948,7 +7948,18 @@ databaseChangeLog:
 #
 # ADVICE:
 #
-# 1) Run ./bin/lint-migrations-file.sh to run core.spec checks against any changes you make here. Liquibase is pretty
+# 1) Think through some of these points when writing a migration, learn from our past mistakes:
+#    - If you need data migration, do you really need to? Could the feature work without it? Prefer not to if possible.
+#      Data in the wild can be vastly different from what you expect, and it's hard to get right.
+#    - If you still need to, make sure it works when Metabase is encrypted. Fields like `metabase_database.details`
+#      or `setting.value` can be encrypted, so you need to decrypt them in your migration. See #42617, #44048.
+#    - Prefer SQL migrations over custom migrations.
+#    - Postgres and MySQL have their differences. Make sure your migration works on both.
+#    - Custom migrations demand a high level of reliability. They should only use code from the standard library. Requiring code
+#      from the Metabase codebase is a bad idea because it can change and break your migration.
+#    - Custom migrations shouldn't use Toucan models; work directly with the table names.
```

Comment:
I add that custom migration tests should also be stable and have few dependencies. Because if the migration depends on something, and the migration test also depends on the same thing, and then that thing changes, the test can pass as if nothing happened but the migration's behaviour actually changes.

Reply:
Isn't that superseded by the custom migrations shouldn't use external code? I don't want to make it like that we allow exceptions.

Also I want to make this list as succinct as possible.

Reply:
I'm suggesting it might be worth pointing out tests separately, because I sometimes see "external code" being used in migration tests. It only takes a couple of extra words:

```diff
- Custom migrations demand a high level of reliability.
+ Custom migrations and their tests demand a high level of reliability.
```

Also, maybe we could link to the "rules" for custom migrations at the top of the namespace here:
https://github.com/metabase/metabase/blob/6899f3cccb34a4c3cb8152e8bb3a3a60c677fa66/src/metabase/db/custom_migrations.clj#L2-L9

Reply:
ok I'll bite. updated